### PR TITLE
fix: LG/LU editing performance optimize

### PR DESCRIPTION
**Pull request by zhixzhan**
## Description
Optimize LG/LU editing performance by:
1. LG/LU parser side delta-parsing
2. Composer side refine utils methods, decrease parsing invoke as low frequency as possible.

LU update time is 1/5 ~ 1/10 compared to previous, LG remain no obvious gain but data flow are optimized.

## Task Item
#minor

## Screenshots
Current LG whole-file & inline editing:
Untitled1
Current LU whole-file & inline editing:
Untitled2

Previous LG whole-file & inline editing:
Untitled3

Previous LU whole-file & inline editing:
Untitled4

### CODE REVIEW

**Comment by boydc2014** [comment]
This looks good, as expected new LU parser brought better performance.

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 1]
```diff
@@ -155,39 +154,53 @@ export const lgDispatcher = () => {
     }) => {
       const { snapshot } = callbackHelpers;
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.updateTemplate(id, content, templateName, template, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(`lg file ${id} not exist`);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:159]:
```suggestion
        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
```

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 2]
```diff
@@ -155,39 +154,53 @@ export const lgDispatcher = () => {
...
   const createLgTemplate = useRecoilCallback(
     (callbackHelpers: CallbackInterface) => async ({ id, template }: { id: string; template: LgTemplate }) => {
       const { snapshot } = callbackHelpers;
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.addTemplate(id, content, template, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(`lg file ${id} not exist`);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:172]:
```suggestion
        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
```

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 3]
```diff
@@ -155,39 +154,53 @@ export const lgDispatcher = () => {
...
   const removeLgTemplate = useRecoilCallback(
     (callbackHelpers: CallbackInterface) => async ({ id, templateName }: { id: string; templateName: string }) => {
       const { snapshot } = callbackHelpers;
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.removeTemplate(id, content, templateName, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(`lg file ${id} not exist`);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:186]:
```suggestion
        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
```

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 4]
```diff
@@ -155,39 +154,53 @@ export const lgDispatcher = () => {
...
   const removeLgTemplates = useRecoilCallback(
     (callbackHelpers: CallbackInterface) => async ({ id, templateNames }: { id: string; templateNames: string[] }) => {
       const { snapshot } = callbackHelpers;
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.removeTemplates(id, content, templateNames, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(`lg file ${id} not exist`);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:200]:
```suggestion
        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
```

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 5]
```diff
@@ -203,9 +216,12 @@ export const lgDispatcher = () => {
     }) => {
       const { snapshot } = callbackHelpers;
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.copyTemplate(id, content, fromTemplateName, toTemplateName, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(`lg file ${id} not exist`);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:221]:
```suggestion
        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
```

File: Composer/packages/client/src/recoilModel/dispatchers/lg.ts [hunk 6]
```diff
@@ -155,39 +155,53 @@ export const lgDispatcher = () => {
...
       const lgFiles = await snapshot.getPromise(lgFilesState);
-      let content = lgFiles.find((file) => file.id === id)?.content ?? '';
-      content = lgUtil.updateTemplate(id, content, templateName, template, lgFileResolver(lgFiles)).content;
-      await updateLgFileState(callbackHelpers, { id, content });
+      const lgFile = lgFiles.find((file) => file.id === id);
+      if (!lgFile) {
+        throw new Error(formatMessage('lg file {id} does not exist.', { id }));
+      }
+      const updatedFile = lgUtil.updateTemplate(lgFile, templateName, template, lgFileResolver(lgFiles));
+      await updateLgFileState(callbackHelpers, { id, content: updatedFile.content, updatedFile });
```
Comment (lei9444) [inline Composer/packages/client/src/recoilModel/dispatchers/lg.ts:163]: I think we can use undatedFile to replace the old one. The updateLgFileState will call worker to parse the file content again. So the async can be removed and we can use set function directly.

**Pull request merged**